File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

vector<int> arr[1005];

bool vis[1005];
void dfs(int cur) {
    vis[cur] = true;
    cout << cur << " ";
    for (auto nxt : arr[cur]) {
        if (vis[nxt] == true) continue ;
        dfs(nxt);
    }
}

queue<int> que;
void bfs() {
    while (!que.empty()) {
        int cur = que.front();
        que.pop();
        cout << cur << " ";
        for (auto nxt : arr[cur]) {
            if (vis[nxt] == true) continue;
            que.push(nxt);
            vis[nxt] = true;
        }
    }
}

int main(void) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m, v;
    cin >> n >> m >> v;

    //그래프 입력받기
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        arr[a].push_back(b);
        arr[b].push_back(a);
    }

    for (int i = 1; i <= n; i++) {
        sort(arr[i].begin(), arr[i].end());
    }

    //DFS
    fill(vis, vis + n + 1, false);
    dfs(v);
    cout << '\n';

    //BFS
    fill(vis, vis + n + 1, false);
    que.push(v);
    vis[v] = true;
    bfs();

    return 0;
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {
	static int N, M, V;
	static List<Integer>[] arr;
	static boolean[] vis;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		V = Integer.parseInt(st.nextToken());
		
		arr = new ArrayList[N + 1];
		for (int i = 0; i <= N; i++) {
			arr[i] = new ArrayList<>();
		}
		
		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int from = Integer.parseInt(st.nextToken());
			int to = Integer.parseInt(st.nextToken());
			arr[from].add(to);
			arr[to].add(from);
		}
		
		StringBuilder sb = new StringBuilder();
		
		// 정점 번호가 작은 것부터 방문하려면 정렬 필요
		for (int i = 1; i <= N; i++) {
		    Collections.sort(arr[i]);
		}

		
		vis = new boolean[N + 1];
		Arrays.fill(vis, false);
		
		//dfs
		dfs(V, sb);
		
		sb.append("\n");
		
		Arrays.fill(vis, false);
		vis[V] = true;
		Queue<Integer> queue = new LinkedList<>();
		queue.add(V);
		
		while(!queue.isEmpty()) {
			int cur = queue.poll();
			sb.append(cur).append(" ");
			for (int nxt : arr[cur]) {
				if (vis[nxt]) continue;
				vis[nxt]= true;
				queue.add(nxt);
			}
		}
		
		
		System.out.println(sb);
		
	}
	
	static void dfs(int cur, StringBuilder sb) {
		vis[cur] = true;
		sb.append(cur).append(" ");
		for (int nxt : arr[cur]) {
			if (!vis[nxt]) {
				dfs(nxt, sb);
			}
		}
	}
}
